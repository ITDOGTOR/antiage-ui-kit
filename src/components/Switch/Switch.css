.ui-kit-switch-container {
	display: flex;
	align-items: center;
	width: fit-content;

	column-gap: 10px;
}

.ui-kit-switch-status-container {
	flex-direction: row-reverse;
	justify-content: center;

	padding: 4px 4px 4px 14px;

	color: var(--base-strong-lighter);
	font-weight: var(--fw-medium);
	font-size: var(--fs-xs);
	line-height: var(--lh-xs);

	background-color: var(--faint-weak);
	border-radius: var(--brad-base);

	cursor: pointer;

	transition: all 0.3s cubic-bezier(0.645, 0.045, 0.355, 1);

	user-select: none;
	touch-action: manipulation;
}

.ui-kit-switch-status-disabled {
	color: var(--faint-strong);

	cursor: not-allowed;
}

.ui-kit-switch-status-checked {
	background-color: var(--complement-weak-lighter);
}

.ui-kit-switch-label {
	cursor: pointer;

	user-select: none;
	touch-action: manipulation;
}

.ui-kit-switch-status-label {
	margin-left: auto;
}

.ui-kit-switch-status-label-checked {
	color: var(--complement-strong-darker);
	font-weight: var(--fw-semibold);
}

.ui-kit-switch {
	min-width: 44px;
	height: 26px;

	font-size: var(--fs-xs);
	line-height: var(--lh-xs);
}

.ui-kit-switch-status-container .ui-kit-switch {
	margin-left: auto;
}

.ui-kit-switch:focus {
	outline: 0;
	box-shadow: 0 0 0 2px var(--faint-weak);
}

.ui-kit-switch.white {
	background-color: var(--faint-weak-darker);
}

.ui-kit-switch.gray {
	background-color: var(--back-faint-weak-darker);
}

.ui-kit-switch .ant-switch-handle {
	width: 22px;
	height: 22px;
}

.ui-kit-switch .ant-switch-handle::before {
	border-radius: 50%;
	box-shadow: 0 4px 4px rgb(0 0 0 / 0.07);
}

.ui-kit-switch.ant-switch-checked {
	background-color: var(--complement-strong);
}

.ui-kit-switch.ant-switch-checked:focus {
	box-shadow: 0 0 0 2px var(--complement-weak);
}

.ui-kit-switch.ant-switch-checked:focus:hover {
	box-shadow: none;
}

.ui-kit-switch .ant-click-animating-node {
	--antd-wave-shadow-color: var(--complement-strong);
}

.ui-kit-switch.ant-switch-checked .ant-switch-handle {
	left: calc(100% - 24px);
}

.ui-kit-switch .ant-switch-loading-icon {
	top: calc(50% - 6px);

	color: var(--back-faint-weak-darker);
}

.ui-kit-switch.ant-switch-checked .ant-switch-loading-icon {
	color: var(--complement-strong);
}

.ui-kit-switch:not(.ant-switch-disabled):active .ant-switch-handle::before {
	right: -10%;
}

.ui-kit-switch:not(.ant-switch-disabled):active.ant-switch-checked .ant-switch-handle::before {
	left: -10%;
}
