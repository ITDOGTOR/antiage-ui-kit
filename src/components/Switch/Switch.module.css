.label {
	position: absolute;
	top: 4px;
	right: 0;
	bottom: 0;
	left: 60px;

	font-size: 14px;

	cursor: pointer;
}

.switch-handle {
	display: none;
}

.switch-handle:disabled + .slider-white,
.switch-handle:disabled + .slider-gray {
	cursor: not-allowed;
	opacity: 0.4;
}

.label-disabled {
	position: absolute;
	top: 4px;
	right: 0;
	bottom: 0;
	left: 60px;

	font-size: 14px;

	cursor: not-allowed;
	opacity: 0.4;
}

.switch {
	position: relative;

	display: inline-block;
	align-items: center;
	width: 44px;
	height: 26px;
}

.slider-white,
.slider-gray {
	position: absolute;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;

	background-color: var(--complement-strong);
	cursor: pointer;

	transition: 0.4s;
}

.slider-white::before,
.slider-gray::before {
	position: absolute;
	bottom: 2px;
	left: 2px;

	width: 22px;
	height: 22px;

	background-color: var(--base-weak);
	box-shadow: 0 4px 4px rgb(0 0 0 / 0.07);

	transition: 0.4s;

	content: "";
}

.slider-white:active::before,
.slider-gray:active::before {
	transform: scaleX(1.2);
	transform-origin: center left;
}

.switch-handle:checked + .slider-white:active::before,
.switch-handle:checked + .slider-gray:active::before {
	transform: scaleX(1.2);
	transform-origin: center right;
}

.switch-handle:checked + .slider-white {
	background-color: var(--faint-weak-up);
}

.switch-handle:checked + .slider-gray {
	background-color: var(--back-faint-middle);
}

.switch-handle:checked + .slider-white::before,
.switch-handle:checked + .slider-gray::before {
	left: 18px;
}

.round {
	border-radius: 13px;
}

.round::before {
	border-radius: 50%;
}
