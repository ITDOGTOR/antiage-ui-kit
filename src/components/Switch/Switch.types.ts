import {SwitchProps as SwitchPropsAntd} from 'antd';
import {ReactNode} from 'react';

import {Theme} from '../index.types';

export declare type SwitchProps = {
	className?: string;
	classNameContainer?: string;
	status?: boolean;
	theme?: Theme;
	label?: ReactNode;
} & SwitchPropsAntd;
