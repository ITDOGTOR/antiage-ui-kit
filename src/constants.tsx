import React from 'react';

export const IS_DISABLED = true;
export const CLASS_NAME = 'classname';
export const CHILD = 'child';
export const THEME = 'white' || 'gray';
export const CHILDREN = (
	<>
		<h1>title</h1>
		<p>description</p>
		<button type="button">button</button>
	</>
);
