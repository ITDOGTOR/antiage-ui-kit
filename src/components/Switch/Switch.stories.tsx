import {ComponentMeta, ComponentStory} from '@storybook/react';
import React from 'react';

import Switch from './Switch';
import {SwitchProps} from './Switch.types';

export default {
	title: 'Switch',
	component: Switch,
	argTypes: {
		theme: {
			defaultValue: 'white',
			table: {defaultValue: {summary: 'white'}},
			control: {type: 'select'},
		},
		status: {
			defaultValue: false,
			control: {type: 'boolean'},
		},
		checked: {
			control: {type: 'boolean'},
		},
		disabled: {
			name: 'disabled',
			description: 'Ant-Design prop',
			defaultValue: false,
			control: {type: 'boolean'},
		},
		label: {control: {disable: true}},
		className: {table: {disable: true}},
		classNameContainer: {table: {disable: true}},
	},
} as ComponentMeta<typeof Switch>;

// eslint-disable-next-line react/function-component-definition
const Template: ComponentStory<typeof Switch> = (args: SwitchProps) => {
	return <Switch {...args} />;
};

export const Sandbox = Template.bind({});

export const Label = Template.bind({});
Label.args = {label: 'Я согласен на обработку персональных данных!'};

export const Status = Template.bind({});
Status.args = {status: true, checkedChildren: 'Добавлен к филиалу', unCheckedChildren: 'Добавить к филиалу'};
