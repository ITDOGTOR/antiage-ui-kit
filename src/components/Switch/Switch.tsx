import classNames from 'classnames/bind';
import React, {ChangeEvent, useState} from 'react';

import styles from './Switch.module.css';
import {SwitchProps} from './Switch.types';

const cx = classNames.bind(styles);

function Switch({
	className,
	theme = 'white',
	children,
	onChange,
	disabled = false,
	defaultChecked = false,
}: SwitchProps) {
	const [inputChecked, setInputChecked] = useState<boolean>(defaultChecked || false);

	const onChangeChecked = (evt: ChangeEvent) => {
		setInputChecked((evt.target as HTMLInputElement).checked);
		onChange?.((evt.target as HTMLInputElement).checked);
	};

	return (
		<label className={cx(styles.switch, theme, className)}>
			<input
				checked={inputChecked}
				className={cx(styles.input)}
				data-testid="switch-input"
				disabled={disabled}
				type="checkbox"
				onChange={onChangeChecked}
			/>

			<span className={cx(styles.switchHandle)} data-testid="switch-label">
				<span className={cx(styles.circle)} />
			</span>

			{children && <span className={cx(styles.label)}>{children}</span>}
		</label>
	);
}

export default Switch;
