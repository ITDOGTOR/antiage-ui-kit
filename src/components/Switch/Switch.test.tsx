import {render, screen} from '@testing-library/react';
import React from 'react';

import {CHILD, CLASS_NAME, IS_DISABLED, THEME} from '../../constants';
import Switch from './Switch';

describe('Switch', () => {
	test('Should render without children', () => {
		render(<Switch onChange={() => {}}>{CHILD}</Switch>);

		expect(screen.getByText('child')).toBeInTheDocument();
	});

	test('Should render with className', () => {
		const switchComponent = render(<Switch className={CLASS_NAME} onChange={() => {}} />);
		expect(switchComponent).toMatchSnapshot();
	});

	test('Must be inactive', () => {
		const switchComponent = render(<Switch data-testid="Switch" disabled={IS_DISABLED} onChange={() => {}} />);

		expect(switchComponent).toMatchSnapshot();
	});

	test('Must be active', () => {
		const switchComponent = render(<Switch data-testid="Switch" disabled={!IS_DISABLED} onChange={() => {}} />);

		expect(switchComponent).toMatchSnapshot();
	});

	test('Should be displayed with a different theme', () => {
		const switchComponent = render(<Switch data-testid="Switch" theme={THEME} onChange={() => {}} />);

		expect(switchComponent).toMatchSnapshot();
	});
});
