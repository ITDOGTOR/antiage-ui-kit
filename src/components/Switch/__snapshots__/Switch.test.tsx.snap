// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Must be active 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="switch"
        for="Switch"
      >
        <input
          class="switch-handle"
          id="Switch"
          type="checkbox"
        />
        <span
          class="slider-white round"
        />
        <span
          class="label"
        />
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="switch"
      for="Switch"
    >
      <input
        class="switch-handle"
        id="Switch"
        type="checkbox"
      />
      <span
        class="slider-white round"
      />
      <span
        class="label"
      />
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Switch Must be inactive 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="switch"
        for="Switch"
      >
        <input
          class="switch-handle"
          disabled=""
          id="Switch"
          type="checkbox"
        />
        <span
          class="slider-white round"
        />
        <span
          class="label-disabled"
        />
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="switch"
      for="Switch"
    >
      <input
        class="switch-handle"
        disabled=""
        id="Switch"
        type="checkbox"
      />
      <span
        class="slider-white round"
      />
      <span
        class="label-disabled"
      />
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Switch Should be displayed with a different theme 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="switch"
        for="Switch"
      >
        <input
          class="switch-handle"
          id="Switch"
          type="checkbox"
        />
        <span
          class="slider-gray round"
        />
        <span
          class="label"
        />
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="switch"
      for="Switch"
    >
      <input
        class="switch-handle"
        id="Switch"
        type="checkbox"
      />
      <span
        class="slider-gray round"
      />
      <span
        class="label"
      />
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Switch Should render with className 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="switch classname"
        for="Switch"
      >
        <input
          class="switch-handle"
          id="Switch"
          type="checkbox"
        />
        <span
          class="slider-white round"
        />
        <span
          class="label"
        />
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="switch classname"
      for="Switch"
    >
      <input
        class="switch-handle"
        id="Switch"
        type="checkbox"
      />
      <span
        class="slider-white round"
      />
      <span
        class="label"
      />
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
